db/changes_test.go

```diff
@@ -286,6 +287,58 @@ func TestDocDeletionFromChannelCoalesced(t *testing.T) {
 	printChanges(changes)
 }
 
+func TestActiveOnlyCacheUpdate(t *testing.T) {
+
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyChanges, base.KeyCache)()
+	log.Printf("Starting actual test code")
+	// Create 10 documents
+	revId := ""
+	var err error
+	for i := 1; i <= 10; i++ {
+		key := fmt.Sprintf("%s_%d", t.Name(), i)
+		body := Body{"foo": "bar"}
+		revId, _, err = db.Put(key, body)
+		require.NoError(t, err, "Couldn't create document")
+	}
+
+	// Tombstone 5 documents
+	for i := 2; i <= 6; i++ {
+		key := fmt.Sprintf("%s_%d", t.Name(), i)
+		_, err = db.DeleteDoc(key, revId)
+		require.NoError(t, err, "Couldn't delete document")
+	}
+
+	waitErr := db.WaitForPendingChanges(context.Background())
+	assert.NoError(t, waitErr)
+
+	changesOptions := ChangesOptions{
+		Since:      SequenceID{Seq: 0},
+		ActiveOnly: true,
+	}
+
+	initQueryCount, _ := base.GetExpvarAsInt("syncGateway_changeCache", "view_queries")
+
+	// Get changes with active_only=true
+	activeChanges, err := db.GetChanges(base.SetOf("*"), changesOptions)
+	require.NoError(t, err, "Error getting changes with active_only true")
+	require.Equal(t, 5, len(activeChanges))
+
+	// Ensure the test is triggering a query, and not serving from DCP-generated cache
+	postChangesQueryCount, _ := base.GetExpvarAsInt("syncGateway_changeCache", "view_queries")
+	assert.Equal(t, postChangesQueryCount, initQueryCount+1)
+
+	// Get changes with active_only=false
+	changesOptions.ActiveOnly = false
+	allChanges, err := db.GetChanges(base.SetOf("*"), changesOptions)
+	require.NoError(t, err, "Error getting changes with active_only true")
+	require.Equal(t, 10, len(allChanges))
+
```

Comment:
This test isn't failing when I remove the fix when run with views (because the view contains tombstones, and is able to populate the cache correctly).

We could add another query count assertion after the last changes request, because we're discarding ActiveOnly=true views, even though they're complete.

For additional coverage to make sure the cache is still being populated for ActiveOnly=false, we could run another changes request at the end and verify the second one didn't issue a query.

```suggestion

	// Ensure the test is triggering another query, and not serving from a cache populated by active_only=true
	postChangesQueryCount, _ = base.GetExpvarAsInt("syncGateway_changeCache", "view_queries")
	assert.Equal(t, initQueryCount+2, postChangesQueryCount)

	// Get changes with active_only=false again, and verify it's served from the cache
	allChanges, err = db.GetChanges(base.SetOf("*"), changesOptions)
	require.NoError(t, err, "Error getting changes with active_only true")
	require.Equal(t, 10, len(allChanges))

	postChangesQueryCount, _ = base.GetExpvarAsInt("syncGateway_changeCache", "view_queries")
	assert.Equal(t, initQueryCount+2, postChangesQueryCount)
```

Reply:
I think the verification of cache population for the activeOnly=false case is handled by other tests, but it wouldn't hurt to duplicate that here.  I'll add that, and review whether it makes sense to add the additional query count check.  I had that at one point and removed it because of the difference between view and n1ql behaviour, but I think you're right that it should be consistent with the current fix.